### ISSUE TITLE
Allow to get multiple coordinates of vectors at the same time 

### ISSUE THREAD

**Issue by Nolkaloid** [godotengine/godot-proposals#727]
**Describe the project you are working on:**
3D top down shooter

**Describe the problem or limitation you are having in your project:**
Operations on different vectors become rather cumbersome when working separately on coordinates:
```gdscript
velocity = Vector3(6, 1, 6)
velocity.x *= 2
velocity.z *= 2
```
**Describe the feature / enhancement and how it helps to overcome the problem or limitation:**
It would be really useful to be able to access multiple coordinates at the same time like in GLSL:
```gdscript
velocity = Vector3(6, 1, 6)
velocity.xz *= 2
```

**Comment by Xrayez** [comment]
You can look at my [`VectorResource` plugin](https://github.com/Xrayez/godot-vector-resource) ([AssetLib](https://godotengine.org/asset-library/asset/502) link, but I recommend downloading from GitHub as it contains some recent fixes), see "vector swizzling syntax" in the examples section.

It doesn't support overloaded operators though (nor GDScript: godotengine/godot#23488), so the way you show it in the snippet (namely `*=`, only `=` operator is supported being the `_set` method) is not completely possible, but it's indeed possible to get any combination of a vector (not sure about the performance though). The vector used internally is always `Vector3` because it can contain `Vector2` representation within itself quite nicely.

Of course it's equally cumbersome if you need to perform this kind of operation frequently enough, because you'd basically have to instantiate an entire `Resource` just for the vector math:
```gdscript
var velocity = VectorResource.new()
velocity.xyz = Vector3(6, 1, 6)
velocity.xz = velocity.xz * 2
```
So see for yourself, this has mostly editing/storage purpose.

The plugin originates from #397.

**Comment by Nolkaloid** [comment]
Thanks for linking the plugin! It's great that there is at least a way to achieve "vector swizzling syntax" right now. But it's again rather cumbersome to use a plugin for something like that. My proposal is just a "small improvement" to gdScript and it's just a question of "quality of life".

**Comment by Xrayez** [comment]
Yeah I'm just showcasing the hurdles as a plugin writer. Having this implemented in core would certainly make plugin development much easier in this regard.

**Comment by bojidar-bg** [comment]
Related to https://github.com/godotengine/godot/issues/6861.

**Comment by Calinou** [comment]
How difficult would this be to implement? Would it just be a matter of adding `xz`, `xy` and `yz` properties to Vector3 with setters and getters?

**Comment by bojidar-bg** [comment]
@Calinou In essence, yes.
This get more contrived since we should probably add `zx`, `yx` and `zy` as well.
Then, the swizzles `xzy`, `yzx`, `yxz`, `zxy`, and `zyx` are a natural extension of it.
Then on Vector2, we would want to have `yx` for feature parity.
And for converting a vector2 into a vector3 we might also want to have `xyo`, `xoy`, `oxy`, `yxo`, `yox`, and `oyx`.
(for feature parity, it might be worth adding swizzles with `o` to vector3 as well, but those are around 18)

**Comment by Calinou** [comment]
@bojidar-bg Are swizzles featuring all components really useful though

**Comment by bojidar-bg** [comment]
Indeed, I can't think of a use for three-component swizzles right now.

Maybe someone else can suggest some usecases for those? :smiley:

**Comment by Nolkaloid** [comment]
It could be useful if you want to change the orientation of your frame. Let's say I have 
`var =Vector3(1, 4, 1)`
and then I change it's orientation like this : 
`var = var.xzy`

**Comment by rcorre** [comment]
What about:

```
enum Component {
    ZERO,
    AXIS_X,
    AXIS_Y,
    AXIS_Z
}

Vector2 Vector3::swizzle2( Component x, Component y );
Vector3 Vector3::swizzle3( Component x, Component y, Component z );
```

It is slightly clumsier than `.xyz()`, but still adds a fair bit of convenience without exploding the API with 27 new properties :)

**Comment by ShivamMukherjee** [comment]
Could this be done using a `union` access to the Vector's components and using an overloaded access(`[]`) operator/new function?
```gd
var vec3: Vector3
var vec2: Vector2
vec3[X] = 1
vec3[Y, Z] = 2
vec2 = vec3[Z, Y]
```
Then introducing the syntactic sugar `vec3.zy` for this new operator/function? I have seen rorre's implementation and it felt quite cumbersome in comparison.

**Comment by rcorre** [comment]
@ShivamMukherjee I believe that would be a more significant change, as:

- gdscript doesn't have a multi-element indexing operator (that I'm aware of)
- What are `X` and `Y` in your example? In that context they will be interpreted as variables, not symbols. Unless we add `X` `Y` `Z` constants to the global namespace, it would be more like `vec3[AXIS_Y, AXIS_Z]`

**Comment by ShivamMukherjee** [comment]
> @ShivamMukherjee I believe that would be a more significant change, as:
> 
> * gdscript doesn't have a multi-element indexing operator (that I'm aware of)
> * What are `X` and `Y` in your example? In that context they will be interpreted as variables, not symbols. Unless we add `X` `Y` `Z` constants to the global namespace, it would be more like `vec3[Vector3.AXIS_Y, Vector3.AXIS_Z]`

They're constants basically. The ordering of the constants will matter.

**Comment by aaronfranke** [comment]
For reference, here are links to PRs discussing this same topic: https://github.com/godotengine/godot/pull/39674 https://github.com/godotengine/godot/pull/51165 https://github.com/godotengine/godot/pull/55220 https://github.com/godotengine/godot/pull/93012

My take is that because converting is usually not as simple as just getting multiple coordinates in a swizzle (ex: converting between 2D and 3D you need to negate the Y axis and scale by 100, or whatever pixels-per-meter your project is using), and due to the inevitable combinatorial explosion, the engine should not provide these functions/properties.

**Comment by Calinou** [comment]
> the engine should not provide these functions/properties.

I agree this would be better implemented at a GDScript level, although I wonder how this would be done in C#. This is likely how GLSL compilers handle it internally (maybe look at Mesa for answers 🙂).

**Comment by aaronfranke** [comment]
@Calinou In C#, this feature is needed even less because C# has a feature called extension methods. These are defined in users' own code, saving the engine from bloat, but behave the same way as a built-in method, providing a seamless user experience. For example:

```cs
// Define this anywhere in your project.
public static Vector2 to_2d(this Vector3 v) {
	return new Vector2(v.x, -v.y) * 100.0f;
}

// Then use it like this:
Vector2 v2 = v3.to_2d();
```

**Comment by Nitbandier** [comment]
for starters having just the three `xy`, `yz` and `xz` would be nice and we can discuss the other swizzles later.

or we can get a function with 27 enums,like so:
```gdscript
    var vector_three := vector3.ONE
    vector_three.get_axis(vector3.AXIS_XY)
```
or a function that get a string like `get_axis("XY")` or someting like these,
but even so i suggest the `xy`,`yz` and `xz` still be added as properties to the vector3 as i think these are more usefull.

**Comment by Joy-less** [comment]
Hi, I support the idea of Vector3 "swizzling" and here's how it could work.

Adding each combination of swizzles to `Vector3`/`Vector3I`/`Vector4`/`Vector4I` (e.g. `xy`, `zyx`, ,`xwzy`) would lead to [144 new properties](https://github.com/godotengine/godot-proposals/issues/12620#issuecomment-2978751793) even without repeated axes, which is too much. So I suggest the following.

Firstly you have just `xy`, `xz` and `yz` for `Vector3` and `Vector3I` because they would be used the most:
```gdscript
var horizontal_player_position: Vector2 = player.position.xz
```
```
Vector3.xy    # Justification: For side-view platformers and maps directly to Vector2
Vector3.xz    # Justification: For top-down games or comparing horizontally
Vector3.yz    # Justification: For forward-platformers

Vector3I.xy
Vector3I.xz
Vector3I.yz
```
```cs
struct Vector3 {
    public Vector2 XY { get; }
    public Vector2 XZ { get; }
    public Vector2 YZ { get; }
}
struct Vector3I {
    public Vector2I XY { get; }
    public Vector2I XZ { get; }
    public Vector2I YZ { get; }
}
```

Then for other combinations, you have multiple-indexers:
```gdscript
var horizontal_player_position: Vector2 = player.position[0, 2]
var horizontal_player_position: Vector2 = player.position[Vector3.Axis.AXIS_X, Vector3.Axis.AXIS_Z]
```
```
Vector3[Vector3.Axis]
Vector3[Vector3.Axis, Vector3.Axis]
Vector3[Vector3.Axis, Vector3.Axis, Vector3.Axis]

Vector3I[Vector3I.Axis]
Vector3I[Vector3I.Axis, Vector3I.Axis]
Vector3I[Vector3I.Axis, Vector3I.Axis, Vector3I.Axis]

Vector4[Vector4.Axis]
Vector4[Vector4.Axis, Vector4.Axis]
Vector4[Vector4.Axis, Vector4.Axis, Vector4.Axis]
Vector4[Vector4.Axis, Vector4.Axis, Vector4.Axis, Vector4.Axis]

Vector4I[Vector4I.Axis]
Vector4I[Vector4I.Axis, Vector4I.Axis]
Vector4I[Vector4I.Axis, Vector4I.Axis, Vector4I.Axis]
Vector4I[Vector4I.Axis, Vector4I.Axis, Vector4I.Axis, Vector4I.Axis]
```
```cs
struct Vector3 {
    public readonly float this[Vector3.Axis index] {
        get => index switch {
            Vector3.Axis.X => X,
            Vector3.Axis.Y => Y,
            Vector3.Axis.Z => Z,
            _ => throw new ArgumentOutOfRangeException(nameof(index)),
        };
    }
    public readonly Vector2 this[Vector3.Axis index1, Vector3.Axis index2] {
        get => new(this[index1], this[index2]);
    }
    public readonly Vector3 this[Vector3.Axis index1, Vector3.Axis index2, Vector3.Axis index3] {
        get => new(this[index1], this[index2], this[index3]);
    }
}
```

**Comment by AThousandShips** [comment]
That would only be relevant for C# though, this isn't C# specific and adding support for multiple indexing to GDScript would be a major change (also I'd say it's unintuitive regardless)

**Comment by Joy-less** [comment]
> @AThousandShips
> That would only be relevant for C# though, this isn't C# specific and adding support for multiple indexing to GDScript would be a major change (also I'd say it's unintuitive regardless)

I had not thought about the fact that GDScript doesn't have method overloading or multiple indexing. You could use named methods instead:
```gdscript
var horizontal_player_position: Vector2 = player.position.to_vec2(0, 2)
```

I'm not sure that multiple indexing would be unintuitive though, since it would be an extension of what's already possible:
```gdscript
var x: float = player.position[0]           # already exists
var xz: Vector2 = player.position[0, 2]     # new feature
```

**Comment by Meorge** [comment]
I think it'd also be worth considering the diminishing returns of the swizzling syntax as it has to get more verbose. If the way you get an XYX version of a Vector3 is with
```gdscript
var new_vec = existing_vec.swizzle(Vector3.AXIS_X, Vector3.AXIS_Y, Vector3.AXIS_X)
```
then that's even more verbose than just
```gdscript
var new_vec = Vector3(existing_vec.x, existing_vec.y, existing_vec.x)
```

Personally, I think a method that takes a string would probably be the best compromise:
```gdscript
var new_vec = existing_vec.swizzle("xyx")
```
Although this has a downside of not being guaranteed to return a specific type (if the string is `"xy"` it'd return a Vector2, if it's `"xyx"` it'd return a Vector3, etc).

**Comment by Joy-less** [comment]
@Meorge I agree using the enums that way is verbose. Although you can use the numbers directly which is shorter:
```gdscript
var new_vec = existing_vec.swizzle(0, 1, 0)
```
Also, one of the main benefits of "swizzling" is you don't have to repeat the expression:
```gdscript
var new_vec = Vector2(game.get_node("entities").get_node("player").global_position.x, game.get_node("entities").get_node("player").global_position.z)
```
```gdscript
var player_global_position = game.get_node("entities").get_node("player").global_position
var new_vec = Vector2(player_global_position.x, player_global_position.z)
```
```gdscript
var new_vec = game.get_node("entities").get_node("player").global_position.swizzle(0, 2)
```

One way that could also work is using separate strings:
```gdscript
var new_vec = existing_vec.swizzle("x", "y", "x")
```

I still much prefer the `xy`, `xz` and `yz` properties though:
```gdscript
var new_vec = game.get_node("entities").get_node("player").global_position.xz
```

**Comment by Meorge** [comment]
I agree the properties would be best. Using the indices doesn't feel particularly great to me, because for the most part the axes are referred to as X, Y, Z, W and not 0, 1, 2, 3. If they could be mapped to an enum then that would be better I suppose, but I imagine that'd cause a lot of naming conflicts.

As for the properties, I don't know how flexible they are in other languages, but intuitively it feels to me like they should be able to support whatever weird combination of axes a user might possibly want to use. If `vec.xz` works, then `vec.zxyx` should work, because the expectation the former example established is that you can list out axes as a property and get the corresponding vector back. But if we don't have a property for each possible combination, the user may be confused why swizzling "suddenly" isn't working.

Rather than being added as "proper" properties with getters and setters, could the swizzle pseudo-properties be added to the vectors' `_get()` methods? Similar to how `RectangleShape2D` implements the compatibility `extents` property?

https://github.com/godotengine/godot/blob/46c495ca21f40f57a7fb9c7cde6143738f1652d4/scene/resources/2d/rectangle_shape_2d.cpp#L50-L58

This would keep it from polluting the auto-generated documentation, so we can manually document it instead. However, I recognize the example I'm pulling from here is also a deprecated compatibility patch, so perhaps it's not best practice for a new feature 😅

**Comment by Joy-less** [comment]
> As for the properties, I don't know how flexible they are in other languages, but intuitively it feels to me like they should be able to support whatever weird combination of axes a user might possibly want to use. If vec.xz works, then vec.zxyx should work, because the expectation the former example established is that you can list out axes as a property and get the corresponding vector back. But if we don't have a property for each possible combination, the user may be confused why swizzling "suddenly" isn't working.

In order to hard-code every swizzle, you would need the following number of properties:
- `Vector2`/`Vector2I` = (2^2) + (2^3) + (2^4) = 28 properties
- `Vector3`/`Vector3I` = (3^2) + (3^3) + (3^4) = 117 properties
- `Vector4`/`Vector4I` = (4^2) + (4^3) + (4^4) = 336 properties
- Total = (28\*2) + (117\*2) + (336\*2) = **962 properties!**

So, clearly not reasonable for non-GDScript languages.

**Comment by Joy-less** [comment]
One thing that could work for C# is to use multiple-indexers with chars.

```cs
Vector2 horizontalPlayerPosition = player.Position['x', 'z'];
```
```cs
struct Vector3 {
    public readonly real_t this[char axisChar] {
        get {
            real_t axisValue = axisChar switch {
                'x' or 'X' => X,
                'y' or 'y' => Y,
                'z' or 'Z' => Z,
            }
            return axisValue;
        }
        set {
            switch (axisChar) {
                case 'x' or 'X':
                    X = value;
                    break;
                case 'y' or 'Y':
                    Y = value;
                    break;
                case 'z' or 'Z':
                    Z = value;
                    break;
            }
        }
    }
    // ...
}
```

Alternatively, would it be possible to source-generate extension methods? For example:
```cs
Vector2 horizontalPlayerPosition = player.Position.XZ(); // Method is generated on-demand
```

**Comment by teatov** [comment]
Hi! I also would really like to see swizzling, and after reading the discussions I have some suggestions.

I agree that implementing it as individual properties (i.e. `some_vec3.zyx`) would bring hundreds of unnecessary members cluttering the Vector APIs (I don't even want to imagine the code editor autocompletion box for those), and implementing a whole new operator (i.e. `some_vec3[z, y, x]`) or a different new language feature is an overkill for this single small purpose.

**So my idea looks like this:**
```gdscript
Vector3(1, 2, 3).swizzle_3(ZYX)      # returns Vector3(3, 2, 1)
Vector2(1, 0).swizzle_4(YXYX)        # returns Vector4(0, 1, 0, 1)
Vector4(2, 2, 2, -1).swizzle_2(WW)   # returns Vector2(-1, -1)

Vector2i(0, 3).swizzle_3(XZY)        # returns Vector3i(0, 0, 3)
Vector2i(0, 3).swizzle_3(XZY, 5)     # returns Vector3i(0, 5, 3)
Vector2i(0, 3).swizzle_4(XZYW, 5, 4) # returns Vector3i(0, 5, 3, 4)
```
#### Details
Let's have separate `@GlobalScope` enums for the permutations of `XYZW`, each for a different number of components, with their bits acting as mappings:
```cpp
// the mapping representation can be something like this:
// from: xyzw xyzw xyzw xyzw
//       0000 0000 0000 0000
//   to: xxxx yyyy zzzz wwww

// so the bits  0010    0100    1000
// mean       x:__z_  y:_y__  z:x___
// so zyx -> xyz

enum Swizzle2 {
	//from:xyzw xyzw
	//  to:xxxx yyyy  // xxxx yyyy
	XX = 0b1000'1000, // x___ x___
	XY = 0b1000'0100, // x___ _y__
	XZ = 0b1000'0010, // x___ __z_
	XW = 0b1000'0001, // x___ ___w
	YX = 0b0100'1000, // _y__ x___
	YY = 0b0100'0100, // _y__ _y__
	YZ = 0b0100'0010, // _y__ __z_
	YW = 0b0100'0001, // _y__ ___w
	ZX = 0b0010'1000, // __z_ x___
	ZY = 0b0010'0100, // __z_ _y__
	ZZ = 0b0010'0010, // __z_ __z_
	ZW = 0b0010'0001, // __z_ ___w
	WX = 0b0001'1000, // ___w x___
	WY = 0b0001'0100, // ___w _y__
	WZ = 0b0001'0010, // ___w __z_
	WW = 0b0001'0001  // ___w ___w
};

enum Swizzle3 {
	// from:xyzw xyzw xyzw
	//   to:xxxx yyyy zzzz  // xxxx yyyy zzzz
	XXX = 0b1000'1000'1000, // x___ x___ x___
	XXY = 0b1000'1000'0100, // x___ x___ _y__
	XXZ = 0b1000'1000'0010, // x___ x___ __z_
	// ...
	WWZ = 0b0001'0001'0010, // ___w ___w __z_
	WWW = 0b0001'0001'0001  // ___w ___w ___w
};

enum Swizzle4 {
	//  from:xyzw xyzw xyzw xyzw
	//    to:xxxx yyyy zzzz wwww  // xxxx yyyy zzzz wwww
	XXXX = 0b1000'1000'1000'1000, // x___ x___ x___ x___
	XXXY = 0b1000'1000'1000'0100, // x___ x___ x___ _y__
	XXXZ = 0b1000'1000'1000'0010, // x___ x___ x___ __z_
	// ...
	// yeah...
	// ...
	WWWZ = 0b0001'0001'0001'0010, // ___w ___w ___w __z_
	WWWW = 0b0001'0001'0001'0001  // ___w ___w ___w ___w
};

// the actual order of bits can be reversed so bit-shifting would be easier.
// this one was picked to be more human-readable.
```

Therefore the amount of constants in an enum for K components would be the number of **K-permutations of 4 with repetition**, so `4^K`.

The total number of new constants would be `4^2 + 4^3 + 4^4 = 16 + 64 + 265 =`***`336`***, which is much smaller that other proposed implementations I've seen (i.e. creating a property for every possible permutation of every size for each Vector type).

That's still a lot, but I would rather have loads of constants in `@GlobalScope` than loads of properties on a variable. Especially since we already have `Key` which has 193 constants.

---
As for the methods, I suggest having separate methods for different sizes, so we would have consistent **editor type hints** (that's also the main reason I chose to group permutations by size):
```gdscript
Vector3(1, 2, 3).swizzle_3(ZYX)     # returns Vector3(3, 2, 1)
Vector2(1, 0).swizzle_4(YXYX)       # returns Vector4(0, 1, 0, 1)
Vector4(2, 2, 2, -1).swizzle_2(WW)  # returns Vector2(-1, -1)
Vector3i(0, 3, 5).swizzle_2(XY)     # returns Vector2i(0, 3)
```
I think the enum keys should be without prefixes, for more elegancy.

Now, we have to consider what happens when swizzling with components that aren't present in a lower-dimensional Vector (i.e. `Vector2(1, 2).swizzle_3(XYW)`). That's more tricky. I have several ideas:
```gdscript
# 1. either have one extra argument for any missing vector component:
Vector2(1, 2).swizzle_3(XYW, 3)     # returns Vector3(1, 2, 3)
#                            ^ any missing component
# 2. or have an argument for each extra component:
Vector2(1, 2).swizzle_3(XYW, 0, 3)  # returns Vector3(1, 2, 3)
#                            ^  ^ for w
#                            ^ for z
# 3. or maybe don't have any extra arguments and just return ZERO
Vector2(1, 2).swizzle_3(XYW) # returns Vector3.ZERO as the argument is "invalid"
```
I would prefer option 2, as it gives more flexibility. Also with options 1 and 2 you wouldn't need to have an extra `O` component for empty values (see [this](https://github.com/godotengine/godot-proposals/issues/727#issuecomment-617376848)). And with option 2 you could do this:
```gdscript
Vector2(2, 3).swizzle_4(ZXYW, 1, 4) # returns Vector4(1, 2, 3, 4)
```
These extra arguments would have default values of `0`, as it's the most common use case, from what I've read in the discussions.

Then, converting from `Vector2` to `Vector3` and back (probably the biggest reason people request this feature in the first place, since swizzling extrapolates from that) would be as simple as:
```gdscript
Vector2(4, 2).swizzle_3(XYZ)   # returns Vector3(4, 2, 0)
Vector3(2, 5, 6).swizzle_2(XY) # returns Vector2(2, 5)
```

---
And as for the actual implementation, you should probably use a mapping with a reversed order of `to` components, rather than the one I've shown above:
```cpp
// from: xyzw xyzw xyzw xyzw
//       0000 0000 0000 0000
//   to: wwww zzzz yyyy xxxx
```
This way if we pass a permutation of a different size than expected, it still works:
```gdscript
Vector4(0, 1, 2, 3).swizzle_2(XY)
# would work the same as
Vector4(0, 2, 3, 4).swizzle_2(WWXY)
# since
# wwww zzzz yyyy xxxx
# 0000 0000 0100 1000
# 0001 0001 0100 1000
#           ^^^^ ^^^^ this is the same
```
Also you can reverse the order of the `from` components for consistency:
```cpp
// from: wzyx wzyx wzyx wzyx
//       0000 0000 0000 0000
//   to: wwww zzzz yyyy xxxx
```
That's the version I used in my implementation.

Here I tried making a very-very-very rough proof-of-concept version of the `Vector2.swizzle_2` method:
```cpp
const uint16_t SWIZZ_SHIFT_LEN = 4;

const uint16_t SWIZZ_SHIFT_X = 0 * SWIZZ_SHIFT_LEN;
const uint16_t SWIZZ_SHIFT_Y = 1 * SWIZZ_SHIFT_LEN;
const uint16_t SWIZZ_SHIFT_Z = 2 * SWIZZ_SHIFT_LEN;
const uint16_t SWIZZ_SHIFT_W = 3 * SWIZZ_SHIFT_LEN;

const uint16_t SWIZZ_FROM_X = 0b0001;
const uint16_t SWIZZ_FROM_Y = 0b0010;
const uint16_t SWIZZ_FROM_Z = 0b0100;
const uint16_t SWIZZ_FROM_W = 0b1000;

const uint16_t SWIZZ_TO_X = 0xF << SWIZZ_SHIFT_X;
const uint16_t SWIZZ_TO_Y = 0xF << SWIZZ_SHIFT_Y;
const uint16_t SWIZZ_TO_Z = 0xF << SWIZZ_SHIFT_Z;
const uint16_t SWIZZ_TO_W = 0xF << SWIZZ_SHIFT_W;

Vector2 Vector2::swizzle_2(Swizzle2 p_swizzle, real_t p_z, real_t p_w) const {
	real_t to_x = Math::NaN;
	real_t to_y = Math::NaN;

	if (p_swizzle & SWIZZ_TO_X) {
		uint16_t from = p_swizzle >> SWIZZ_SHIFT_X;
		if (from & SWIZZ_FROM_X) {
			to_x = x;
		} else if (from & SWIZZ_FROM_Y) {
			to_x = y;
		} else if (from & SWIZZ_FROM_Z) {
			to_x = p_z;
		} else if (from & SWIZZ_FROM_W) {
			to_x = p_w;
		}
	}

	if (p_swizzle & SWIZZ_TO_Y) {
		uint16_t from = p_swizzle >> SWIZZ_SHIFT_Y;
		if (from & SWIZZ_FROM_X) {
			to_y = x;
		} else if (from & SWIZZ_FROM_Y) {
			to_y = y;
		} else if (from & SWIZZ_FROM_Z) {
			to_y = p_z;
		} else if (from & SWIZZ_FROM_W) {
			to_y = p_w;
		}
	}

	// should also discuss what to do here.
	// are there any Vector values that can be used as "error" values?
	// Vector2.INF maybe?
	ERR_FAIL_COND_V_MSG(Math::is_nan(to_x) || Math::is_nan(to_y), Vector2(),
			vformat("Received invalid swizzle mapping:  %s.", String::num_int64(p_swizzle, 2)));

	return Vector2(to_x, to_y);
}
```
I compiled it and it works.

For making all the `Swizzle3` and `Swizzle4` permutations probably code generation should be used. Maybe a little Python script, which can be put in `misc/scripts` for future use if needed.

Also the enums themselves could be stored in a separate `.inc` file. In my implementation I placed it in `core/math/vector_swizzle_enums.inc`. This way you could just put them there and never look at them again. The same could be done for `BIND_CORE_ENUM_CONSTANT` macros in `core/core_constants.cpp`, but I don't know how radical of a pattern break that is.

Aaand also all those constants should probably be hidden from the docs, or else the `@GlobalScope` page will be a nightmare. Only a description of the enums themselves may be left saying something like "Contains all N-permutations with repetition of XYZW".

---
It's far from perfect, and I surely have missed something, so I'm open to discussions. If all goes fine, I'd be glad to contribute it myself!

Also I didn't consider swizzling assignment (`some_vec3.zxy = other_vec3`) because I believe that this is enough:
```gdscript
some_vec3 = other_vec2.swizzle_3(XZY)
# instead of
some_vec3 = Vector3(other_vec3.y, other_vec3.z, other_vec3.x)
```
For me the ***actual purpose of swizzling is that I don't have to type the same thing multiple times*** (or create extra vars just for reassignment). I would have to edit the variable name in just one place. And this achieves that.

And one of most important things is, this doesn't add any new language features, so it won't be a massive headache (hopefully).

**Comment by Calinou** [comment]
> It's far from perfect, and I surely have missed something, so I'm open to discussions. If all goes fine, I'd be glad to contribute it myself!

I appreciate the effort about not needing extra members or language features, but this ends up being a lot more writing than GLSL-style swizzling (while not being as discoverable). While code is read a lot more often than it is written, swizzling happens to be one of those use cases where it's beneficial to write less characters (so you can fit more things per line).

If having hundreds of members listed in the documentation or autocompletion is a problem, we can special-case them to not show up in the class reference and autocompletion.

**Comment by Joy-less** [comment]
If the Godot team is willing to add hundreds of extra properties/methods per language, I think a good approach would be:
- Add `get_swizzle(String) -> Variant` and `set_swizzle(String, Variant) -> void` methods. Optionally add generic wrappers.
- For languages where properties could be hidden (GDScript and C#, but not C++ to my knowledge), also define properties that wrap around or provide direct implementations for them.

Example of `get_swizzle` and `set_swizzle` method implementations:
```cs
public Variant GetSwizzle(string combination) {
    float At(int index) {
        return combination[index] switch {
            'X' or 'x' => X,
            'Y' or 'y' => Y,
            'Z' or 'z' => Z,
        };
    }
    return combination.Length switch {
        1 => At(0),
        2 => new Vector2(At(0), At(1)),
        3 => new Vector3(At(0), At(1), At(2)),
        4 => new Vector4(At(0), At(1), At(2), At(3)),
    };
}
```
```cs
public void SetSwizzle(string combination, Variant value) {
    float SetAt(int index, float value) {
        switch (combination[index]) {
            case 'X' or 'x': X = value; break;
            case 'Y' or 'y': Y = value; break;
            case 'Z' or 'z': Z = value; break;
            default: throw new Exception();
        }
    }
    
    if (combination.Length == 1) {
        float typedValue = (float)value;
        SetAt(0, typedValue);
    }
    else if (combination.Length == 2) {
        Vector2 typedValue = (Vector2)value;
        SetAt(0, typedValue[0]);
        SetAt(1, typedValue[1]);
    }
    else if (combination.Length == 3) {
        Vector3 typedValue = (Vector3)value;
        SetAt(0, typedValue[0]);
        SetAt(1, typedValue[1]);
        SetAt(2, typedValue[2]);
    }
    else if (combination.Length == 4) {
        Vector4 typedValue = (Vector4)value;
        SetAt(0, typedValue[0]);
        SetAt(1, typedValue[1]);
        SetAt(2, typedValue[2]);
        SetAt(3, typedValue[3]);
    }
    else {
        throw new Exception();
    }
}
```

Example of static property implementations:
```cs
[EditorBrowsable(EditorBrowsableState.Never)]
public Vector2 XZ {
    get => (Vector2)GetSwizzle("XZ");
    set => SetSwizzle("XZ", value);
}
```
```cs
[EditorBrowsable(EditorBrowsableState.Never)]
public Vector2 XZ {
    get => new Vector2(X, Z);
    set {
        X = value.X;
        Y = value.Y;
    }
}
```

Result:
```cs
Vector3 position = new(1, 1, 1);

Vector2 flatPosition = (Vector2)position.GetSwizzle("XZ");
position.SetSwizzle("XZ", new Vector2(0, 0));

Vector2 flatPosition = position.XZ;
position.XZ = new Vector2(0, 0);
```

Language-specific alternatives:
- For GDScript, this could be a language feature.
- For C#, these could be extension methods (but would need brackets and can't be assigned - maybe the extension properties feature from .NET 10.0 could be useful here?).

**Comment by coldays** [comment]
I created a pull request for swizzling in C# with source generation instead of hardcoding the properties.
The properties are not EditorBrowsable and not DebuggerBrowsable to reduce noise.
I made the Vector structs partial in order to add the properties instead of doing extension properties since GodotSharp is not yet at .NET 10.
That PR only generates getters, but can also be extended to generate setters as well.
